### PR TITLE
dialects: (builtin) add IntegerAttr truncation and use in individual rewrite

### PR DESCRIPTION
**Pull request by alexarice**
`AdditionOfSameVariablesToMultiplyByTwo` has a corner case where the bitwidth is 1. This small change fixes this issue.

### CODE REVIEW

**Comment by alexarice** [comment]
@superlopuh should this be dealt with by normalization?

**Comment by superlopuh** [comment]
This is a classic, I believe CC @math-fehr . I think this particular case will be dealt with with normalization

**Comment by superlopuh** [comment]
It would be good to have a filecheck for this

**Review by math-fehr — COMMENTED** [review]
Question, why was the previous implementation wrong?
With an `i1`, the constant 2 should be folded to `0` right?
So `x + x` gets folded to `x * 0`, which is correct AFAIK?

**Comment by alexarice** [comment]
> Question, why was the previous implementation wrong? With an `i1`, the constant 2 should be folded to `0` right? So `x + x` gets folded to `x * 0`, which is correct AFAIK?

It doesn't get converted to 0, it just adds an `i1` with `2` inside which then fails a verify check

**Comment by alexarice** [comment]
@math-fehr @superlopuh any idea how to proceed here? Would it be better to always wrap integers to their range, or at least have an IntegerAttr initialiser that does this?

File: xdsl/transforms/individual_rewrite.py [hunk 1]
```diff
@@ -17,10 +17,11 @@ class AdditionOfSameVariablesToMultiplyByTwo(RewritePattern):
...
-            assert isinstance(op.lhs.type, IntegerType | IndexType)
+            assert isinstance(type := op.lhs.type, IntegerType | IndexType)
+            value = 0 if isinstance(type, IntegerType) and type.bitwidth <= 1 else 2
             rewriter.replace_matched_op(
                 [
-                    li_op := arith.ConstantOp(IntegerAttr(2, op.lhs.type)),
+                    li_op := arith.ConstantOp(IntegerAttr(value, type)),
                     arith.MuliOp(op.lhs, li_op),
                 ]
             )
```
Comment (superlopuh) [inline xdsl/transforms/individual_rewrite.py:27]: isn't the whole thing just 0 if the bitwidth is 1? So could branch earlier and explicitly have it as a special case

Comment (alexarice) [reply xdsl/transforms/individual_rewrite.py:27]: Could do

**Comment by superlopuh** [comment]
Another possible solution is to add a truncating constructor to accept any size of input

**Comment by alexarice** [comment]
This is what I meant by having a wrapping IntegerAttr initialiser

**Comment by math-fehr** [comment]
I would agree with the wrapping IntegerAttr init

**Comment by alexarice** [comment]
I've pushed a version which adds an option to truncate an IntegerAttr. If this seems reasonable I can potentially split that into a separate PR or at least add some unit tests.

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -397,21 +397,26 @@ def verify_value(self, value: int):
...
         """
         min_value, max_value = self.value_range()
         if not (min_value <= value.data < max_value):
-            return None
+            if not truncate_bits:
+                return None
+            value = IntAttr(value.data & (2**self.bitwidth - 1))
 
-        if self.signedness.data == Signedness.SIGNLESS:
+        if self.signedness.data != Signedness.UNSIGNED:
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:419]: why this change?

**Review by superlopuh — COMMENTED** [review]
I'm not sure that these are entirely composable things, and signedness complicates things. I'm not sure how to best do this in the general case, handling all possible signednesses and scenarios. Would it make sense to change this PR to only handle the i1 case in the pattern, and then start a separate discussion on how to truncate integers?

File: xdsl/dialects/builtin.py (hunk 2 above)
Comment (alexarice) [reply xdsl/dialects/builtin.py:419]: Doing the bitwise and seems to wrap everything into the range [0,2**bitwidth), so for the signed case we need to normalise it back to the correct range. This seemed like the easiest way to do this but it could be refactored

**Comment by alexarice** [comment]
I don't really see how this isn't composable into 
1. Add optional truncation for IntegerAttr (with added tests)
2. Enable the flag on this test.

I am also not sure I understand the concerns around signedness

**Review by superlopuh — COMMENTED** [review]
I played around with the code locally, and I get how it works a bit better now, happy to go with this direction. Could you please add unit tests for the helper, and a filechec test for the canonicalization pattern?

**Comment by alexarice** [comment]
I've realised I can probably replace the `& (2**8 - 1)` with `% 2**8`, which acts exactly the same as far as I can tell and is way more readable

File: tests/filecheck/transforms/individual_rewrite/add-same.mlir [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+%1 = arith.addi %0, %0 : i32
+
+// -----
+
+// CHECK:      %[[#var:]] = "test.op"() : () -> i1
+// CHECK-NEXT: %[[#zero:]] = arith.constant false
+// CHECK-NEXT: %{{.*}} = arith.muli %[[#var]], %[[#zero]] : i1
+
+%0 = "test.op"() : () -> (i1)
+%1 = arith.addi %0, %0 : i1
```
Comment (superlopuh) [inline tests/filecheck/transforms/individual_rewrite/add-same.mlir:18]: nit: are the regexes here really more readable than giving names to ssa values and checking for those directly?

**Review by superlopuh — APPROVED** [review]

File: tests/filecheck/transforms/individual_rewrite/add-same.mlir (hunk 3 above)
Comment (alexarice) [reply tests/filecheck/transforms/individual_rewrite/add-same.mlir]: I've removed 1, but can't/haven't named the constants

Comment (superlopuh) [reply tests/filecheck/transforms/individual_rewrite/add-same.mlir]: If you name the inputs, the constants they are replaced with inherit their hints, but this is more of a nit at this point

File: xdsl/dialects/builtin.py [hunk 4]
```diff
@@ -397,21 +397,26 @@ def verify_value(self, value: int):
                 f"values in the range [{min_value}, {max_value})"
             )
 
-    def normalized_value(self, value: IntAttr) -> IntAttr | None:
+    def normalized_value(
+        self, value: IntAttr, *, truncate_bits: bool = False
```
Comment (math-fehr) [inline xdsl/dialects/builtin.py:401]: I think a name like `wrapping` would be clearer than `truncate_bits`.

**Review by math-fehr — APPROVED** [review]
Nice! Looks good to me!

File: xdsl/dialects/builtin.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/dialects/builtin.py:401]: Hmm, what's it wrapping?

Comment (math-fehr) [reply xdsl/dialects/builtin.py:401]: Yes, and truncating has already a meaning in LLVM/arith, which is to go from let's say i32 to i16, which is not really what we are doing here

Comment (superlopuh) [reply xdsl/dialects/builtin.py:401]: But wrapping is a feature of the operation, not of a value normalization step. I'm happy for an addition to be wrapping, but it seems a bit out of context here.

Comment (superlopuh) [reply xdsl/dialects/builtin.py:401]: Here's the relevant API that I have most experience with https://developer.apple.com/documentation/swift/int/init(truncating:)

File: tests/filecheck/transforms/individual_rewrite/add-same.mlir (hunk 3 above)
Comment (alexarice) [reply tests/filecheck/transforms/individual_rewrite/add-same.mlir]: I don't want the test to be checking that the constant has the same name as the original operation, and so would prefer to use the regex

File: xdsl/dialects/builtin.py (hunk 4 above)
Comment (alexarice) [reply xdsl/dialects/builtin.py:401]: Did we come to a conclusion?

Comment (superlopuh) [reply xdsl/dialects/builtin.py:401]: We can hash it out on the zulip if you want, I'm happy to be outvoted if you both think wrapping is more intuitive here
